2단계 미션 제출합니다.

owenoweny opened the pull request:
안녕하세요, 개인 사정으로 PR이 너무 늦어졌습니다 죄송합니다 ㅠ
이번 리뷰 잘 부탁드립니다.

Hue9010 reviewed (changes requested):
안녕하세요. 이번 리뷰를 맞은 휴입니다!
고민을 많이 하신게 보이는데 그만큼 훌륭히 잘 해주셨어요. 👍👍👍
부분적으로 아쉬운 곳들 피드백 남겼는데 확인하시고 궁금한 건 언제든 질문 주세요. 😉

kukaro commented:
안녕하십니까 신입 공채 4기 Coeat 입니다.

상세한 피드백 감사합니다.
말씀드린 피드백을 바탕으로 다시 수정하였습니다.
다시 리뷰 부탁드립니다.
항상 감사합니다.

Hue9010 commented on src/test/java/domain/CookiesTest.java:
```diff
@@ -0,0 +1,49 @@
+                .contains("logined=true");
+    }
+
+    @DisplayName("잘못된 cookies(= 두개) 생성자 테스트")
+    @Test
+    void invalidConstructorTest() {
+        String header = "logi=ned=true;dodo=10";
+        assertThatThrownBy(()->{
+            new Cookies(header);
+        }).isInstanceOf(InvaliCookieException.class);
+    }
```
good! 👍
테스트를 만들 때 챙겨야 할 게 성공하는 테스트만 만들지 않는 겁니다.
구현을 하면서도 보통 정상적인 플로우대로 흘러가는 상황만 고려하기 쉽상인데 로직이 수정되다 보면 실패해야 할 게 실패하지 않는 문제가 생길 수도 있어요. 특히나 이런 경우는 문제가 생겼다는 걸 알게되기까지 오래 걸릴 수도 있습니다.

Hue9010 commented on src/main/java/domain/RequestLine.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Map;
+
+public class RequestLine {
+    private final String method;
+    private final String urlPath;
+    private final Map<String, String> queries;
```
version이 생략된 것 같지만 😉  전 지금 이 상태가 딱 좋다고 생각합니다. 버전을 추가하면 생각보다 번거로운 작업들이 생기는데 Http버전 자체가 중요한 내용은 아니라고 봐서 그 외의 것에 집중하는게 더 좋다고 생각해요.
그런면에서 필요한 핵심 사항에 잘 집중 해 주신 것 같아요.👍

Hue9010 commented on src/main/java/controller/LoginController.java:
```diff
@@ -0,0 +1,43 @@
+        User user = DataBase.findUserById(request.getBody("userId"));
+        if(user.isLogin(request.getBody("password"))){
+            Response response = Response.ofRedirect("/index.html");
+            response.addHeader("Set-Cookie", "logined=true; Path=/");
```
쿠키 추가는 직접 헤더의 키를 지정해서 넣지 말고 Response에서 쿠키를 추가하는 메서드를 제공해 줄 수 있을 것 같아요.

지금은 새로 쿠키를 추가할 때마다 기존 쿠키가 사라질텐데 Headers에서 Map에 넣기 전에 기존값을 꺼내뒀다가 새로 요청 온 값과 합친다면 쿠키를 계속 추가해서 넣을 수 있겠네요.

Hue9010 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,55 +1,76 @@
+
+            printResponse(out, response);
+
+        } catch (IOException | URISyntaxException e) {
             logger.error(e.getMessage());
         }
```
Q. 혹시 이전 단계들을 진행하면서 ControllerAdvice를 사용하여 Exception을 핸들링 해보셨나요? 그렇다면 이번 과제에서도 비슷한 기능을 제공한다면 어떻게 할 수 있을 까요? (단순한 형태로) 🤔
[Spring Handle Exception](https://jaehun2841.github.io/2018/08/30/2018-08-25-spring-mvc-handle-exception/#%EB%93%A4%EC%96%B4%EA%B0%80%EB%A9%B0)

A. 코멘트를 이쪽에 남겨드리는 이유가 있겠죠? 😉 정답은 없지만 개인적으로 봤을 때 간단히 구현할 수 있는게 이곳 같아 보여요.
logger.error로 끝내는게 아니라 특정 예외를 잡아 BadRequest, NotFound 등 각 예외에 맞는 응답을 해 줄 수 있겠죠.

kukaro replied:
아 맞네요! 고민하고 있었는데 여기서 하면됬었네요. 감사합니다!

Hue9010 reviewed (approved):
빠르게 잘 반영해주셨네요. 👍 
이번 단계는 머지하도록 할게요. 간단한 내용 피드백 남겼는데 확인 후 궁금한건 언제든 질문 주세요.
다음 단계도 언제든 pr 보내주시면 피드백 드리도록 할게요. 😉
고생하셨습니다.

owenoweny committed:
feat: 로그인 기능 구현

- HttpCookie 클래스 구현
- user/login api 구현
- http 응답 헤더 추가 메소드 구현

owenoweny committed:
feat: 세션 저장 기능 구현

owenoweny committed:
feat: user/list 조회 기능 구현

- list.html: 템플릿 엔진 적용
- /user/list api 구현
- HttpRequest가 HttpCookie 필드를 가지도록 수정

owenoweny committed:
test: RequestProcessor.saveUser 단위 테스트 구현

liquidjoo commented on src/main/java/db/DataBase.java:
```diff
@@ -18,6 +19,10 @@ public static User findUserById(String userId) {
     }
 
+    public static void deleteAll() {
+        users = new HashMap<>();
```
새로운 객체를 만든다는 느낌보단 비워주는 느낌이 좋을 것 같아요~

liquidjoo commented on src/main/java/webserver/HttpCookie.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Map;
+
+public class HttpCookie {
+    public static final String SESSION_KEY = "JSSESSIONID";
+    private final Map<String, String> values;
```
```suggestion
    public static final String SESSION_KEY = "JSSESSIONID";
    
    private final Map<String, String> values;
```
상수와 상태 값은 구분 지어주면 좋을 것 같아요~

liquidjoo commented on src/main/java/webserver/RequestProcessor.java:
```diff
@@ -36,4 +44,54 @@ public HttpResponse saveUser(HttpRequest httpRequest) {
+    @HandleRequest(path = "/user/login", httpMethod = HttpMethod.POST)
+    public HttpResponse login(HttpRequest httpRequest) {
+        HttpRequestBody body = httpRequest.getBody();
+        //TODO : NPE 예외처리
```
예외처리는 TODO 로 두는 것 보다 코드로 구현을 해보면 어떨까요?

liquidjoo commented on src/main/java/webserver/RequestProcessor.java:
```diff
@@ -36,4 +44,54 @@ public HttpResponse saveUser(HttpRequest httpRequest) {
+    public HttpResponse login(HttpRequest httpRequest) {
+        HttpRequestBody body = httpRequest.getBody();
+        //TODO : NPE 예외처리
+        String userId = body.get("userId");
+        String password = body.get("password");
+
+        User user = DataBase.findUserById(userId);
+        if (user == null || !user.getPassword().equals(password)) {
+            return HttpResponse.found("http://localhost:8080/user/login_failed.html");
+        }
+
+        Session session = SessionManager.createSession();
+        session.addAttribute("userId", userId);
+        session.addAttribute("password", password);
+
+        HttpCookie httpCookie = HttpCookie.of(SESSION_KEY, session.getSessionId());
+        HttpResponse httpResponse = HttpResponse.found(INDEX_REDIRECT_URI);
+        httpResponse.addCookie(httpCookie, "/");
```
여러 책임이 존재하는 것 같아요~
메서드 분리를 통해 책임을 분리해보면 어떨까요?

역할또한 나눌 수 있는 것들이 보이네요~ 역할도 나누어 보면 좋습니다 :)

liquidjoo commented on src/main/java/webserver/Session.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Map;
+
+public class Session {
+    private final String sessionId;
```
원시 값을 포장한 객체를 만들어 보면 어떨까요?
어떤 장점이 있을까요?

liquidjoo commented on src/main/java/webserver/SessionManager.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Map;
+import java.util.UUID;
+
+public class SessionManager {
```
Session 이 계속 만들어지기만 한다면 어떤 일이 발생할 수 있을까요?

liquidjoo commented on src/test/java/webserver/RequestProcessorTest.java:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    @Test
+    void login() {
+    }
+
+    @Test
+    void list() {
+    }
```
해당 테스트 코드 어떤 의미일까요?

liquidjoo commented on src/test/java/webserver/RequestProcessorTest.java:
```diff
@@ -0,0 +1,110 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.*;
+
+class RequestProcessorTest {
```
꼼꼼한 테스트 코트 작성 좋습니다 👍 
하지만 현재 만들어진 클래스에 비해 테스트가 많이 부족해 보여요~
다음 요청 시 테스트 코드 작성해 보면 좋을 것 같아요!

liquidjoo reviewed (changes requested):
안녕하세요 요환님 :)
2단계 미션 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다~ 확인하고 다시 요청해 주세요 🙇
